Test Failure: com.ibm.ws.request.timing.monitor.fat.RequestTimingMbeanTest.testHungServletRequest

Error:

testHungServletRequest:junit.framework.AssertionFailedError: 2021-03-06-07:00:32:252 [Unexpected FFDC reporting java.lang.IllegalStateException was found (count = 1): C:\jazz-build-engines\jazz112.hursley.ibm.com\EBC.PROD.WASRTC\build\dev\image\output\wlp\usr\servers\RequestTimingFeatureServer\logs\ffdc\ffdc_21.03.06_07.00.12.0.log
>------Start of DE processing------ = [06/03/2021, 07:00:12:225 GMT]
>Exception = java.lang.IllegalStateException
>Source = com.ibm.ws.kernel.launch.internal.FrameworkManager
>probeid = 1247
>Stack Dump = java.lang.IllegalStateException: java.lang.IllegalAccessException: class com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl$VirtualMachine cannot access class sun.tools.attach.HotSpotVirtualMachine (in module jdk.attach) because module jdk.attach does not export sun.tools.attach to unnamed module @7dc5e7b4
> at com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl$VirtualMachine.remoteDataDump(HotSpotJavaDumperImpl.java:341)
> at com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl.createThreadDump(HotSpotJavaDumperImpl.java:195)
> at com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl.dump(HotSpotJavaDumperImpl.java:101)
> at com.ibm.ws.kernel.launch.internal.FrameworkManager.dumpJava(FrameworkManager.java:1233)
> at com.ibm.ws.kernel.launch.internal.FrameworkManager.dumpJava(FrameworkManager.java:1219)
> at com.ibm.ws.kernel.launch.internal.LibertyProcessImpl.createJavaDump(LibertyProcessImpl.java:106)
> at com.ibm.ws.request.timing.manager.ThreadDumpManager$1.run(ThreadDumpManager.java:108)
> at com.ibm.ws.threading.internal.SchedulingRunnableFixedHelper.run(SchedulingRunnableFixedHelper.java:290)
> at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
> at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:238)
> at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
> at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
> at java.base/java.lang.Thread.run(Thread.java:831)
>Caused by: java.lang.IllegalAccessException: class com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl$VirtualMachine cannot access class sun.tools.attach.HotSpotVirtualMachine (in module jdk.attach) because module jdk.attach does not export sun.tools.attach to unnamed module @7dc5e7b4
> at java.base/jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:385)
> at java.base/java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:687)
> at com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl$VirtualMachine.remoteDataDump(HotSpotJavaDumperImpl.java:339)
]
at componenttest.custom.junit.runner.FATRunner.blowup(FATRunner.java:303)
at componenttest.custom.junit.runner.FATRunner.access$500(FATRunner.java:65)
at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:253)
at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:343)
at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:177)

Resolution:
For Java 16 we need to add the --illegal-access=permit jvm parameter to avoid this error during a server dump.

--- a/dev/com.ibm.ws.request.timing.monitor_fat/fat/src/com/ibm/ws/request/timing/monitor/fat/RequestTimingMbeanTest.java
+++ b/dev/com.ibm.ws.request.timing.monitor_fat/fat/src/com/ibm/ws/request/timing/monitor/fat/RequestTimingMbeanTest.java
@@ -44,6 +44,7 @@ import componenttest.annotation.TestServlet;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
 import componenttest.custom.junit.runner.Mode.TestMode;
+import componenttest.topology.impl.JavaInfo;
 import componenttest.topology.impl.LibertyServer;
 
 /**
@@ -59,7 +60,7 @@ import componenttest.topology.impl.LibertyServer;
  * request during the mbean call which happens inside the initial servlet call.
  */
 @RunWith(FATRunner.class)
-@SkipForRepeat({"MPM23", "MPM22", "MPM20"})
+@SkipForRepeat({ "MPM23", "MPM22", "MPM20" })
 public class RequestTimingMbeanTest {
 
     private static final Class<RequestTimingMbeanTest> c = RequestTimingMbeanTest.class;
@@ -88,6 +89,13 @@ public class RequestTimingMbeanTest {
     @BeforeClass
     public static void setUp() throws Exception {
         ShrinkHelper.defaultDropinApp(server, "RequestTimingWebApp", "com.ibm.ws.request.timing.app");
+
+        // Fix for Java 16 Hotspot exception during java dump
+        JavaInfo java = JavaInfo.forCurrentVM();
+        if (java.majorVersion() != 8) {
+            server.copyFileToLibertyServerRoot("illegalAccess/jvm.options");
+        }
+
         server.startServer();
     }
 
@@ -532,7 +540,7 @@ public class RequestTimingMbeanTest {
      * allow a test to finish
      *
      * @param countToWaitFor
-     *                           Value looked for in CountDownLatch
+     *            Value looked for in CountDownLatch
      * @throws Exception
      */
     private void waitInServletForCountDownLatch(int countToWaitFor) throws Exception {
@@ -606,13 +614,13 @@ public class RequestTimingMbeanTest {
      * test
      *
      * @param th
-     *                            -- array of threads
+     *            -- array of threads
      * @param numReqs
-     *                            -- number of requests is the number of threads needed
+     *            -- number of requests is the number of threads needed
      * @param servletTestName
-     *                            -- Used for request to servlet
+     *            -- Used for request to servlet
      * @param testMethodName
-     *                            -- Used for printing to logs
+     *            -- Used for printing to logs
      */
     private void createRequestThreads(Thread[] th, int numReqs) {
         // Send N servlet requests to server, last request used to terminate

--- a/dev/com.ibm.ws.request.timing.monitor_fat/publish/files/illegalAccess/jvm.options
+++ b/dev/com.ibm.ws.request.timing.monitor_fat/publish/files/illegalAccess/jvm.options
@@ -1,0 +1,1 @@
+--illegal-access=permit

